src/core/queryCache.ts

```diff
@@ -20,7 +20,46 @@ interface QueryHashMap {
   [hash: string]: Query<any, any>
 }
 
-type QueryCacheListener = (query?: Query) => void
+interface NotifyEventQueryAdded {
+  type: 'queryAdded'
+  query: Query<any, any>
+}
+
+interface NotifyEventQueryRemoved {
+  type: 'queryRemoved'
+  query: Query<any, any>
+}
+
+interface NotifyEventObserverAdded {
+  type: 'observerAdded'
+  query: Query<any, any>
+}
+
+interface NotifyEventObserverRemoved {
+  type: 'observerRemoved'
+  query: Query<any, any>
+}
+
+interface NotifyEventUpdateResults {
+  type: 'updateResults'
+  query: Query<any, any>
+}
+
+interface NotifyEventDispatch {
+  type: 'dispatch'
+  query: Query<any, any>
+  action: Action<any, any>
+}
+
+type QueryCacheNotifyEvent =
+  | NotifyEventQueryAdded
+  | NotifyEventQueryRemoved
+  | NotifyEventObserverAdded
+  | NotifyEventObserverRemoved
+  | NotifyEventUpdateResults
+  | NotifyEventDispatch
```

Comment:
Dispatch seems a bit implementation specific, what about `NotifyEventQueryUpdate` and `NotifyEventQueryObserverUpdate`? Maybe also an idea to add the observer to the observer events?

Reply:
Done and done.